File: sdks/python/rogue_client/websocket.py
"""
WebSocket Client for Rogue Agent Evaluator real-time updates.
"""

import asyncio
import json
import logging
from typing import Optional, Callable, Dict, List, Any
import websockets
from websockets.exceptions import ConnectionClosed, WebSocketException
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from websockets.client import ClientConnection

from .types import WebSocketMessage, WebSocketEventType

logger = logging.getLogger(__name__)


class RogueWebSocketClient:
    """WebSocket client for real-time updates."""

    def __init__(self, base_url: str, job_id: Optional[str] = None):
        self.base_url = base_url.replace("http", "ws").rstrip("/")
        self.job_id = job_id
        self.websocket: Optional["ClientConnection"] = None
        self.event_handlers: Dict[WebSocketEventType, List[Callable]] = {}
        self.is_connected = False
        self.reconnect_attempts = 0
        self.max_reconnect_attempts = 5
        self.reconnect_delay = 1.0
        self._stop_event = asyncio.Event()

    async def connect(self) -> None:
        """Connect to WebSocket."""
        if self.is_connected:
            return

        ws_url = f"{self.base_url}/ws"
        if self.job_id:
            ws_url = f"{self.base_url}/ws/{self.job_id}"

        try:
            self.websocket = await websockets.connect(
                ws_url,
            )  # type: ignore[assignment]
            self.is_connected = True
            self.reconnect_attempts = 0
            self._emit("connected", {"url": ws_url})

            # Start message handling task
            asyncio.create_task(self._handle_messages())

        except Exception as e:
            self._emit("error", {"error": str(e)})
            raise

    async def disconnect(self) -> None:
        """Disconnect from WebSocket."""
        self._stop_event.set()
        self.is_connected = False

        if self.websocket:
            await self.websocket.close()  # type: ignore[attr-defined]
            self.websocket = None

        self._emit("disconnected", {})

    def on(self, event: WebSocketEventType, handler: Callable) -> None:
        """Add event handler."""
        if event not in self.event_handlers:
            self.event_handlers[event] = []
        self.event_handlers[event].append(handler)

    def off(self, event: WebSocketEventType, handler: Callable) -> None:
        """Remove event handler."""
        if event in self.event_handlers:
            try:
                self.event_handlers[event].remove(handler)
            except ValueError:
                pass

    def remove_all_listeners(self, event: Optional[WebSocketEventType] = None) -> None:
        """Remove all event handlers."""
        if event:
            self.event_handlers.pop(event, None)
        else:
            self.event_handlers.clear()

    async def _handle_messages(self) -> None:
        """Handle incoming WebSocket messages."""
        try:
            while not self._stop_event.is_set() and self.websocket:
                try:
                    message_data = await asyncio.wait_for(
                        self.websocket.recv(), timeout=1.0  # type: ignore[attr-defined]
                    )

                    try:
                        message_dict = json.loads(message_data)
                        message = WebSocketMessage(**message_dict)
                        await self._handle_message(message)
                    except (json.JSONDecodeError, ValueError) as e:
                        logger.error(f"Failed to parse WebSocket message: {e}")

                except asyncio.TimeoutError:
                    # Timeout is expected, continue loop
                    continue
                except ConnectionClosed:
                    logger.info("WebSocket connection closed")
                    break
                except WebSocketException as e:
                    logger.error(f"WebSocket error: {e}")
                    break

        except Exception as e:
            logger.error(f"Error in message handler: {e}")
        finally:
            self.is_connected = False
            if self.reconnect_attempts < self.max_reconnect_attempts:
                await self._schedule_reconnect()

    async def _handle_message(self, message: WebSocketMessage) -> None:
        """Handle a parsed WebSocket message."""
        if message.type == "job_update":
            self._emit("job_update", message.data)
        elif message.type == "chat_update":
            self._emit("chat_update", message.data)
        else:
            logger.warning(f"Unknown WebSocket message type: {message.type}")

    def _emit(self, event: WebSocketEventType, data: Any) -> None:
        """Emit event to handlers."""
        handlers = self.event_handlers.get(event, [])
        for handler in handlers:
            try:
                if asyncio.iscoroutinefunction(handler):
                    asyncio.create_task(handler(event, data))
                else:
                    handler(event, data)
            except Exception as e:
                logger.error(f"Error in event handler: {e}")

    async def _schedule_reconnect(self) -> None:
        """Schedule reconnection attempt."""
        if self._stop_event.is_set():
            return

        self.reconnect_attempts += 1
        delay = self.reconnect_delay * (2 ** (self.reconnect_attempts - 1))

        logger.info(
            f"Scheduling reconnect attempt {self.reconnect_attempts} in {delay}s"
        )
        await asyncio.sleep(delay)

        if not self._stop_event.is_set():
            try:
                await self.connect()
            except Exception as e:
                logger.error(f"Reconnection failed: {e}")
                if self.reconnect_attempts < self.max_reconnect_attempts:
                    await self._schedule_reconnect()

File: packages/sdk/src/websocket.ts
/**
 * WebSocket Client for Rogue Agent Evaluator real-time updates
 */

import WebSocket from 'ws';
import {
  WebSocketMessage,
  WebSocketEventType,
  WebSocketEventHandler
} from './types';

export class RogueWebSocketClient {
  private ws?: WebSocket;
  private baseUrl: string;
  private jobId?: string;
  private eventHandlers: Map<WebSocketEventType, WebSocketEventHandler[]> = new Map();
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;
  private isConnecting = false;

  constructor(baseUrl: string, jobId?: string) {
    this.baseUrl = baseUrl.replace(/^http/, 'ws').replace(/\/$/, '');
    this.jobId = jobId;
  }

  /**
   * Connect to WebSocket
   */
  async connect(): Promise<void> {
    if (this.isConnecting || (this.ws && this.ws.readyState === WebSocket.OPEN)) {
      return;
    }

    this.isConnecting = true;

    return new Promise((resolve, reject) => {
      try {
        const wsUrl = this.jobId 
          ? `${this.baseUrl}/ws/${this.jobId}`
          : `${this.baseUrl}/ws`;

        this.ws = new WebSocket(wsUrl);

        this.ws.on('open', () => {
          this.isConnecting = false;
          this.reconnectAttempts = 0;
          this.emit('connected', { url: wsUrl });
          resolve();
        });

        this.ws.on('message', (data: WebSocket.Data) => {
          try {
            const message: WebSocketMessage = JSON.parse(data.toString());
            this.handleMessage(message);
          } catch (error) {
            console.error('Failed to parse WebSocket message:', error);
          }
        });

        this.ws.on('close', (code: number, reason: Buffer) => {
          this.isConnecting = false;
          this.emit('disconnected', { code, reason: reason.toString() });
          
          if (code !== 1000 && this.reconnectAttempts < this.maxReconnectAttempts) {
            this.scheduleReconnect();
          }
        });

        this.ws.on('error', (error: Error) => {
          this.isConnecting = false;
          this.emit('error', { error: error.message });
          reject(error);
        });

      } catch (error) {
        this.isConnecting = false;
        reject(error);
      }
    });
  }

  /**
   * Disconnect from WebSocket
   */
  disconnect(): void {
    if (this.ws) {
      this.ws.close(1000, 'Client disconnect');
      this.ws = undefined;
    }
  }

  /**
   * Add event handler
   */
  on(event: WebSocketEventType, handler: WebSocketEventHandler): void {
    if (!this.eventHandlers.has(event)) {
      this.eventHandlers.set(event, []);
    }
    this.eventHandlers.get(event)!.push(handler);
  }

  /**
   * Remove event handler
   */
  off(event: WebSocketEventType, handler: WebSocketEventHandler): void {
    const handlers = this.eventHandlers.get(event);
    if (handlers) {
      const index = handlers.indexOf(handler);
      if (index > -1) {
        handlers.splice(index, 1);
      }
    }
  }

  /**
   * Remove all event handlers
   */
  removeAllListeners(event?: WebSocketEventType): void {
    if (event) {
      this.eventHandlers.delete(event);
    } else {
      this.eventHandlers.clear();
    }
  }

  /**
   * Check if connected
   */
  get isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }

  /**
   * Handle incoming WebSocket message
   */
  private handleMessage(message: WebSocketMessage): void {
    switch (message.type) {
      case 'job_update':
        this.emit('job_update', message.data);
        break;
      case 'chat_update':
        this.emit('chat_update', message.data);
        break;
      default:
        console.warn('Unknown WebSocket message type:', message.type);
    }
  }

  /**
   * Emit event to handlers
   */
  private emit(event: WebSocketEventType, data: any): void {
    const handlers = this.eventHandlers.get(event);
    if (handlers) {
      handlers.forEach(handler => {
        try {
          handler(event, data);
        } catch (error) {
          console.error('Error in WebSocket event handler:', error);
        }
      });
    }
  }

  /**
   * Schedule reconnection attempt
   */
  private scheduleReconnect(): void {
    this.reconnectAttempts++;
    const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);
    
    setTimeout(() => {
      if (this.reconnectAttempts <= this.maxReconnectAttempts) {
        this.connect().catch(error => {
          console.error('Reconnection failed:', error);
        });
      }
    }, delay);
  }
}
